### DISCUSSION TITLE
Temporary record parameters should always be checked for being temporary

### DISCUSSION THREAD

**Discussion by christophstuber** [StefanMaron/BusinessCentral.LinterCop#732] [New Rule]
When passing a temporary record as a parameter, you almost always specify the parameter to be by var. However, when you do that, the temporary property becomes more of a suggestion. If the variable you pass to the function is not temporary, the parameter will not be temporary either.

This example shows the problem:

Since you normally don't want to use triggers with temporary records, it's almost always DeleteAll() without the RunTrigger or false. This means safety checks for deletes are not run.
```AL
local procedure MyProcedure()
var
  ItemLedgerEntry: Record "Item Ledger Entry";
begin
  MyProcedure2(ItemLedgerEntry);
end;

local procedure MyProcedure2(var TempItemLedgerEntry: Record "Item Ledger Entry" temporary)
begin
  TempItemLedgerEntry.Reset();
  TempItemLedgerEntry.DeleteAll();
end;
```

Checking the caller with static code analysis would be a first step, but the only safe version at runtime (especially for public functions) is to check IsTemporary and either exit or Error(). I prefer Error, as the method is designed to be used only with temporary records.

```AL
local procedure MyProcedure()
var
  ItemLedgerEntry: Record "Item Ledger Entry";
begin
  // Additional warning here
  MyProcedure2(ItemLedgerEntry);
end;

local procedure MyProcedure2(var TempItemLedgerEntry: Record "Item Ledger Entry" temporary)
begin
  // Check for this pattern
  if not TempItemLedgerEntry.IsTemporary() then
    Error('%1 must be temporary', TempItemLedgerEntry.TableCaption());

  TempItemLedgerEntry.Reset();
  TempItemLedgerEntry.DeleteAll();
end;
```

**Comment by jwikman** [comment]
This is a good one. 👍

A (maybe obvious) condition is that the table of the parameter does not have `TableType = Temporary`.
